crates/relayer/src/keyring.rs

```diff
@@ -98,6 +99,40 @@ impl KeyEntry {
             })
         }
     }
+
+    /// Get key from seed file
+    pub fn from_seed_file(key_file_content: &str, hd_path: &HDPath) -> Result<Self, Error> {
+        let key_file = serde_json::from_str(key_file_content).map_err(Error::encode)?;
+        Self::from_key_file(key_file, hd_path)
+    }
+
+    // NOTE: Cosmos does not keep `v` (recovery code) in the `(r, s, v)` tuple
+    // for secp256k1, while Ethereum (and therefore Evmos) does.
+    //
+    // Cosmos implementation: https://github.com/cosmos/cosmos-sdk/blob/main/crypto/keys/secp256k1/secp256k1_cgo.go
+    // Evmos docs: https://docs.evmos.org/modules/evm/04_transactions.html
+    pub fn sign_message(
+        &self,
+        message: &[u8],
+        address_type: &AddressType,
+    ) -> Result<Vec<u8>, Error> {
+        let hashed_message: GenericArray<u8, U32> = match address_type {
+            AddressType::Ethermint { ref pk_type } if pk_type.ends_with(".ethsecp256k1.PubKey") => {
+                keccak256_hash(message).into()
+            }
+            AddressType::Cosmos | AddressType::Ethermint { .. } => Sha256::digest(message),
+        };
+
+        // SAFETY: hashed_message is 32 bytes, as expected in `Message::from_slice`,
+        // so `unwrap` is safe.
+        let message = Message::from_slice(&hashed_message).unwrap();
+
+        // TODO: Call `.sign_ecdsa_recoverable` for Evmos.
```

Comment:
Let's address this TODO before merging

Reply:
@tomtau We would love to hear what you think we should do here?

Reply:
it should be a recoverable signature, see the previous code:
https://github.com/informalsystems/hermes/pull/2863/files#diff-08bcf6496eddb7428a6a3c7894c2ac3e33be5b3767c0a27a1953eda3dbbabe96L391

you need to enable this feature on secp256k1: https://github.com/rust-bitcoin/rust-secp256k1/blob/master/Cargo.toml#LL28

Reply:
if the pubkey ends with `.ethsecp256k1.PubKey`

Reply:
Right I agree the old code calls `sign_ecdsa_recoverable` but the previous line `let (_, sig_bytes) =` throws away the recovery parameter, which is the first element of the tuple.

I [wrote some test code](https://github.com/kevinji/secp256k1-test/blob/main/src/lib.rs) to validate that the remaining bytes of `sign_ecdsa_recoverable` is the same as the bytes from `sign_ecdsa`.

This is why I was asking what should be done for Evmos as I believe the current behavior is already incorrect.

Reply:
Hm, indeed. Unlike Ethereum, Ethermint doesn't seem to recover pubkeys from signatures (but just looks them up for the account), so it should be ok to use `sign_ecdsa`.

Reply:
yeah, ethermint only do recovery for eth tx that's signed by ethereum compatible wallets, but ibc txs go through normal cosmos-sdk ante handler pipelines.

Reply:
Great, I removed the TODO per the conversation here and updated the comment a bit.